readme.md
# Deadlock Crating Integrations For Bob's Mods
Adds 2 tiers to match with Bob's Logistics extra belts, and more closely ties crating machines to their little sisters, the belt boxes.

Updates crafting speeds of existing crating machines to respect the belt overhaul belt speed setting. Updates recipes for crating machines to use belt boxes as ingredients and be a little more balanced for Bob's Mods


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 1.3.0
  Features:
    - Moved to Factorio version 1.1
---------------------------------------------------------------------------------------------------
Version: 1.2.1
  Bugfixes:
    - Fix tier 4 and 5 requiring the wrong belts
---------------------------------------------------------------------------------------------------
Version: 1.2.0
  Features:
    - Update to 0.18
---------------------------------------------------------------------------------------------------
Version: 1.1.4
  Bugfixes:
    - Fixed Crating Machine recipe
---------------------------------------------------------------------------------------------------
Version: 1.1.3
  Bugfixes:
    - Fixed a crash when only vanilla items are available
---------------------------------------------------------------------------------------------------
Version: 1.1.2
  Bugfixes:
    - Fixed a crash without "Bobs Logistics" and "Loader Redux" or "Deadlock Integrations
    - Fixed recipes related to Ultimate Transport Belts
---------------------------------------------------------------------------------------------------
Version: 1.1.1
  Bugfixes:
    - Fix DCM.LOADER_THREE variable
---------------------------------------------------------------------------------------------------
Version: 1.1.0
  Changes:
    - Modernized code base again to use more functions and reduce code duplicacy
    - Make code GPL compliant
---------------------------------------------------------------------------------------------------
Version: 1.0.0
  Changes:
    - Modernized code base to be more aligned with "DeadlockCrating"
    - Fixed crashs with a recent update to "DeadlockCrating"
    - Made "deadlock-integrations" optional
    - Add optional support for "LoaderRedux" Loaders
---------------------------------------------------------------------------------------------------
Version: 0.0.4
  Changes:
    - Corrected tinting of the entity to correctly show the colour when placed
---------------------------------------------------------------------------------------------------
Version: 0.0.3
  Changes:
    - Corrected something so minor didn't even write it here
---------------------------------------------------------------------------------------------------
Version: 0.0.2
  Changes:
    - Changed method of creating the new tiers to better reflect changes in the base Crating Mods
    - Updated code to gracefully handle situations where you aren't using Bob's Logistics but you want faster crating machines anyways
---------------------------------------------------------------------------------------------------
Version: 0.0.1
  Features:
    - Updates recipes to be better balanced with Bobs Mods
    - Adds 2 new tiers to match Bob's logistics


data-updates.lua
require('prototypes.shared')


----------------------------

-- iterate through tiers to add entities
for i=4,DCM.TIERS do
	local machine = table.deepcopy(data.raw["assembling-machine"]["assembling-machine-"..i])
	machine.name = "deadlock-machine-packer-entity-"..i
	machine.icon = nil
	machine.icons = {
		{ icon = "__DeadlockCrating__/graphics/icons/crating-icon-base-128.png" },
		{ icon = "__DeadlockCrating__/graphics/icons/crating-icon-mask-128.png", tint = DCM.TIER_COLOURS[i] },
	}
	machine.minable.result = "deadlock-machine-packer-item-"..i
	machine.crafting_speed = i
	machine.module_specification = {
	  module_info_icon_shift = { 0, 0.8 },
	  module_slots = 1
	}
	machine.allowed_effects = {"consumption"}
	machine.crafting_categories = {"packing"}
	machine.max_health = DCM.get_scale(i, DCM.TIERS, 300, 400)
	local energy_table = DCM.get_energy_table(i, DCM.TIERS, 500, 1000, 0.1, 5 - i)
	machine.energy_usage = energy_table.active
	machine.energy_source = {
		type = "electric",
		usage_priority = "secondary-input",
		drain = energy_table.passive,
		emissions_per_second_per_watt = energy_table.emissions, -- replaces old emissions parameter
	}
	machine.working_sound = { filename = "__DeadlockCrating__/sounds/deadlock-crate-machine.ogg", volume = 0.7 }
	machine.animation = {
		layers = {
			{
				hr_version = {
					draw_as_shadow = true,
					filename = "__DeadlockCrating__/graphics/entities/high/crating-shadow.png",
					animation_speed = 1 / i,
					repeat_count = 60,
					height = 192,
					scale = 0.5,
					shift = {1.5, 0},
					width = 384
				},
				draw_as_shadow = true,
				filename = "__DeadlockCrating__/graphics/entities/low/crating-shadow.png",
				animation_speed = 1 / i,
				repeat_count = 60,
				height = 96,
				scale = 1,
				shift = {1.5, 0},
				width = 192	
			},
			{
				hr_version = {
					filename = "__DeadlockCrating__/graphics/entities/high/crating-base.png",
					animation_speed = 1 / i,
					priority = "high",
					frame_count = 60,
					line_length = 10,
					height = 192,
					scale = 0.5,
					shift = {0, 0},
					width = 192
				},
				filename = "__DeadlockCrating__/graphics/entities/low/crating-base.png",
				animation_speed = 1 / i,
				priority = "high",
				frame_count = 60,
				line_length = 10,
				height = 96,
				scale = 1,
				shift = {0, 0},
				width = 96	
			},
			{
				hr_version = {
					filename = "__DeadlockCrating__/graphics/entities/high/crating-mask.png",
					animation_speed = 1 / i,
					priority = "high",
					repeat_count = 60,
					height = 192,
					scale = 0.5,
					shift = {0, 0},
					width = 192,
					tint = DCM.TIER_COLOURS[i],
				},
				filename = "__DeadlockCrating__/graphics/entities/low/crating-mask.png",
					animation_speed = 1 / i,
				priority = "high",
				repeat_count = 60,
				height = 96,
				scale = 1,
				shift = {0, 0},
				width = 96,	
				tint = DCM.TIER_COLOURS[i],
			}
		}
	}
	machine.working_visualisations = {
		{
		  animation = {
			hr_version = {
			  animation_speed = 1 / i,
			  blend_mode = "additive",
			  filename = "__DeadlockCrating__/graphics/entities/high/crating-working.png",
			  frame_count = 30,
			  line_length = 10,
			  height = 192,
			  priority = "high",
			  scale = 0.5,
			  tint = DCM.brighter_colour(DCM.TIER_COLOURS[i]),
			  width = 192
			},
			animation_speed = 1 / i,
			blend_mode = "additive",
			filename = "__DeadlockCrating__/graphics/entities/low/crating-working.png",
			frame_count = 30,
			line_length = 10,
			height = 96,
			priority = "high",
			tint = DCM.brighter_colour(DCM.TIER_COLOURS[i]),
			width = 96
		  },
		  light = {
			color = DCM.brighter_colour(DCM.TIER_COLOURS[i]),
			intensity = 0.4,
			size = 9,
			shift = {0, 0},
		  },
		},
	}
	data:extend{machine}
end

-- crating machine items
for i=4,DCM.TIERS do
	data:extend {
		{
			type = "item",
			name = "deadlock-machine-packer-item-"..i,
			subgroup = "production-machine",
			stack_size = 50,
			icons = {
				{ icon = "__DeadlockCrating__/graphics/icons/crating-icon-base-"..DCM.ITEM_ICON_SIZE..".png" },
				{ icon = "__DeadlockCrating__/graphics/icons/crating-icon-mask-"..DCM.ITEM_ICON_SIZE..".png", tint = DCM.TIER_COLOURS[i] },
			},
			icon_size = DCM.ITEM_ICON_SIZE,
			order = "z"..i,
			place_result = "deadlock-machine-packer-entity-"..i,
			flags = {},
		}
	}
end


-- RECIPES for the machines
data:extend {
	{
	type = "recipe",
	name = "deadlock-machine-packer-recipe-4",
	enabled = false,
	ingredients = {
        {"deadlock-machine-packer-item-3", 1},
		{"turbo-transport-belt-beltbox",4},
		{"steel-plate",10}
	},
	result = "deadlock-machine-packer-item-4",
	energy_required = 9.0;
	},
	{
	type = "recipe",
	name = "deadlock-machine-packer-recipe-5",
	enabled = false,
	ingredients = {
        {"deadlock-machine-packer-item-4", 1},
        {"ultimate-transport-belt-beltbox",4},
		{"steel-plate",10}
	},
	result = "deadlock-machine-packer-item-5",
	energy_required = 10.0;
	},
 }

 






prototypes/shared.lua
-- Here we need to add the tiers
DCM = {}

-- size of machine icons and crate background
DCM.ITEM_ICON_SIZE = 32

-- how many crates to use up a whole vanilla stack
DCM.STACK_DIVIDER = 5

-- how many tiers of tech?
DCM.TIERS = 5
-- there are kind-of 4 tiers - all imported mod items go in tier 4 - only affects crafting tab
DCM.TAB_TIERS = DCM.TIERS + 1

-- which items can be crated, in which tier
DCM.ITEM_TIER = {
	[4] = {"wood", "iron-ore", "copper-ore", "stone", "coal", "iron-plate", "copper-plate", "steel-plate", "stone-brick"},
	[5] = {
		"copper-cable",
		"iron-gear-wheel",
		"iron-stick",
		"sulfur",
		"plastic-bar",
		"solid-fuel",
		"electronic-circuit",
		"advanced-circuit"
	}
}

DCM.TIER_COLOURS = {
	[1] = {r = 210, g = 180, b = 80},
	[2] = {r = 210, g = 60, b = 60},
	[3] = {r = 80, g = 180, b = 210},
	[4] = {r = 165, g = 10, b = 225},
	[5] = {r = 10, g = 225, b = 25}
}

-- which loaders to use based on bobs logistic
if mods["boblogistics"] and mods["LoaderRedux"] then
	DCM.LOADER_ONE = "loader"
	DCM.LOADER_TWO = "fast-loader"
	DCM.LOADER_THREE = "express-loader"
	DCM.LOADER_FOUR = "purple-loader"
	DCM.LOADER_FIVE = "green-loader"
else
	DCM.LOADER_ONE = "transport-belt-beltbox"
	DCM.LOADER_TWO = "fast-transport-belt-beltbox"
	DCM.LOADER_THREE = "express-transport-beltbox"
	if data.raw.item["ultimate-transport-belt-beltbox"] then
		DCM.LOADER_FOUR = "express-transport-belt"
		DCM.LOADER_FIVE = "express-transport-belt"
	else
		DCM.LOADER_FOUR = "ultimate-transport-belt-beltbox"
		DCM.LOADER_FIVE = "turbo-transport-belt-beltbox"
	end
end

DCM.MACHINE_PREFIX = "deadlock-crating-machine-"
DCM.TECH_PREFIX = "deadlock-crating-"

-- for calculating scales of energy, health etc.
function DCM.get_scale(this_tier, tiers, lowest, highest)
	return lowest + ((highest - lowest) * ((this_tier - 1) / (tiers - 1)))
end

-- Energy and pollution. They just couldn't make it easy, could they
function DCM.get_energy_table(this_tier, tiers, lowest, highest, passive_multiplier, pollution)
	local total = DCM.get_scale(this_tier, tiers, lowest, highest)
	local passive_energy_usage = total * passive_multiplier
	local active_energy_usage = total * (1 - passive_multiplier)
	return {
		passive = passive_energy_usage .. "KW", -- passive energy drain as a string
		active = active_energy_usage .. "KW", -- active energy usage as a string
		emissions = pollution / active_energy_usage / 1000 -- pollution/s/W
	}
end

-- brighter version of tier colour for working vis glow & lights
function DCM.brighter_colour(c)
	local w = 240
	return {r = math.floor((c.r + w) / 2), g = math.floor((c.g + w) / 2), b = math.floor((c.b + w) / 2)}
end


data-final-fixes.lua
local DCM = require "__DeadlockCrating__/prototypes/shared"

DCM.ITEM_TIER = {
	[4] = {
		"wood",
		"iron-ore",
		"copper-ore",
		"stone",
		"coal",
		"iron-plate",
		"copper-plate",
		"steel-plate",
		"stone-brick"
	},
	[5] = {
		"copper-cable",
		"iron-gear-wheel",
		"iron-stick",
		"sulfur",
		"plastic-bar",
		"solid-fuel",
		"electronic-circuit",
		"advanced-circuit"
	}
}
DCM.TIERS = 5
DCM.TIER_COLOURS = {
	[1] = {r = 210, g = 180, b = 80},
	[2] = {r = 210, g = 60, b = 60},
	[3] = {r = 80, g = 180, b = 210},
	[4] = {r = 165, g = 10, b = 225},
	[5] = {r = 10, g = 225, b = 25}
}

if mods["boblogistics"] and mods["LoaderRedux"] then
	DCM.LOADER_ONE = "loader"
	DCM.LOADER_TWO = "fast-loader"
	DCM.LOADER_THREE = "express-loader"
	DCM.LOADER_FOUR = "purple-loader"
	DCM.LOADER_FIVE = "green-loader"
elseif mods["deadlock-integrations"] then
	DCM.LOADER_ONE = "transport-belt-beltbox"
	DCM.LOADER_TWO = "fast-transport-belt-beltbox"
	DCM.LOADER_THREE = "express-transport-belt-beltbox"
	if data.raw.item["ultimate-transport-belt-beltbox"] then
		DCM.LOADER_FOUR = "turbo-transport-belt-beltbox"
		DCM.LOADER_FIVE = "ultimate-transport-belt-beltbox"
	else
		DCM.LOADER_FOUR = "express-transport-belt"
		DCM.LOADER_FIVE = "express-transport-belt"
	end
else
	DCM.LOADER_ONE = "transport-belt"
	DCM.LOADER_TWO = "fast-transport-belt"
	DCM.LOADER_THREE = "express-transport-belt"
	if data.raw.item["ultimate-transport-belt"] then
		DCM.LOADER_FOUR = "turbo-transport-belt"
		DCM.LOADER_FIVE = "ultimate-transport-belt"
	else
		DCM.LOADER_FOUR = "express-transport-belt"
		DCM.LOADER_FIVE = "express-transport-belt"
	end
end

for tier = 1, DCM.TIERS do
	DCM.create_machine_entity(
		tier,
		nil,
		nil,
		nil,
		nil,
		nil,
		(tier < DCM.TIERS) and (DCM.MACHINE_PREFIX .. (tier + 1)) or nil,
		nil
	)
end

for tier = 4, DCM.TIERS do
	DCM.create_machine_item(tier)
end

DCM.create_machine_recipe(
	4,
	{
		{"deadlock-crating-machine-3", 1},
		{DCM.LOADER_FOUR, 4},
		{"steel-plate", 10}
	}
)

DCM.create_machine_recipe(
	5,
	{
		{"deadlock-crating-machine-4", 1},
		{DCM.LOADER_FIVE, 4},
		{"steel-plate", 10}
	}
)

data.raw.recipe[DCM.MACHINE_PREFIX .. "1"].ingredients = {
	{DCM.LOADER_ONE, 8},
	{"assembling-machine-1", 1},
	{"steel-plate", 10}
}

data.raw.recipe[DCM.MACHINE_PREFIX .. "2"].ingredients = {
	{DCM.MACHINE_PREFIX .. "1", 1},
	{DCM.LOADER_TWO, 4},
	{"steel-plate", 10}
}

data.raw.recipe[DCM.MACHINE_PREFIX .. "3"].ingredients = {
	{DCM.MACHINE_PREFIX .. "2", 1},
	{DCM.LOADER_THREE, 4},
	{"steel-plate", 10}
}

function getCraftingSpeed(level)
	return level * settings.startup["bobmods-logistics-beltspeedperlevel"].value / 15
end

if
	settings.startup["bobmods-logistics-beltoverhaulspeed"] and
	settings.startup["bobmods-logistics-beltoverhaulspeed"].value
then
	-- yellow
	if data.raw["assembling-machine"][DCM.MACHINE_PREFIX .. "1"] then
		data.raw["assembling-machine"][DCM.MACHINE_PREFIX .. "1"].crafting_speed = getCraftingSpeed(2)
	end
	-- red
	if data.raw["assembling-machine"][DCM.MACHINE_PREFIX .. "2"] then
		data.raw["assembling-machine"][DCM.MACHINE_PREFIX .. "2"].crafting_speed = getCraftingSpeed(3)
	end
	-- blue
	if data.raw["assembling-machine"][DCM.MACHINE_PREFIX .. "3"] then
		data.raw["assembling-machine"][DCM.MACHINE_PREFIX .. "3"].crafting_speed = getCraftingSpeed(4)
	end
	-- purple
	if data.raw["assembling-machine"][DCM.MACHINE_PREFIX .. "4"] then
		data.raw["assembling-machine"][DCM.MACHINE_PREFIX .. "4"].crafting_speed = getCraftingSpeed(5)
	end
	-- green
	if data.raw["assembling-machine"][DCM.MACHINE_PREFIX .. "5"] then
		data.raw["assembling-machine"][DCM.MACHINE_PREFIX .. "5"].crafting_speed = getCraftingSpeed(6)
	end
end

-- set replace groups
for tier = 1, DCM.TIERS do
	if data.raw["assembling-machine"][DCM.MACHINE_PREFIX .. tier] then
		data.raw["assembling-machine"][DCM.MACHINE_PREFIX .. tier].fast_replaceable_group = "crating-machine"
		if (tier < 5) then
			data.raw["assembling-machine"][DCM.MACHINE_PREFIX .. tier].next_upgrade = DCM.MACHINE_PREFIX .. tier + 1
		else
			data.raw["assembling-machine"][DCM.MACHINE_PREFIX .. tier].next_upgrade = ""
		end
	end
end

unit = table.deepcopy(data.raw.technology["inserter-capacity-bonus-6"].unit)
unit.count = unit.count * 2
DCM.create_crating_technology(4, nil, {DCM.TECH_PREFIX .. 3}, unit)

unit = table.deepcopy(data.raw.technology["inserter-capacity-bonus-7"].unit)
unit.count = unit.count * 2
DCM.create_crating_technology(5, nil, {DCM.TECH_PREFIX .. 4}, unit)
